libraries/AP_BattMonitor/AP_BattMonitor_AD7091R5.cpp

```diff
@@ -0,0 +1,236 @@
+        volt_buff_pt = _volt_pin.get() -1;
+    }
+    if (_curr_pin >= 1 && _curr_pin <= AD7091R5_NO_OF_CHANNELS) {
+        curr_buff_pt = _curr_pin.get() -1;
+    }
+
+    if (_first) { // only the first instance read the i2c device
+        _first = false;
+        _i2c_polling = true;
+        _state.healthy = false;
+        // probe i2c device
+        _dev = hal.i2c_mgr->get_device(AD7091R5_I2C_BUS, AD7091R5_I2C_ADDR);
+        if (_dev) {
+            WITH_SEMAPHORE(_dev->get_semaphore());
+            _dev->set_retries(10); // lots of retries during probe
+            //config
+            if (_reset() && _config()) {
+                _state.healthy = true;
+                _dev->set_retries(2); // drop to 2 retries for runtime
+                _dev->register_periodic_callback(AD7091R5_PERIOD_USEC, FUNCTOR_BIND_MEMBER(&AP_BattMonitor_AD7091R5::_timer, void));
+            }
+        }
+    }
+}
+
+/**
+ * config the adc
+ * - command mode
+ * - use external 3.3 reference
+ * - all channels enabled
+ * - set address pointer register to read the adc results
+ */
+bool AP_BattMonitor_AD7091R5::_config(){
+    uint8_t data[6] = {AD7091R5_CONF_ADDR, reg_conf_h, reg_conf_l, AD7091R5_CHAN_ADDR, reg_chan, AD7091R5_RESULT_ADDR};
+    return _dev->transfer(data, sizeof(data), nullptr, 0);
+}
+
+/**
+ * @brief read - read the voltage and curren
+ * 
+ */
+void AP_BattMonitor_AD7091R5::read(){
+    
+    WITH_SEMAPHORE(sem);
+    //voltage
+    _state.voltage = (_data_to_volt(_analog_data[volt_buff_pt].data) - _volt_offset) * _volt_multiplier;
+
+    //current amps
+    _state.current_amps = (_data_to_volt(_analog_data[curr_buff_pt].data) - _curr_amp_offset) * _curr_amp_per_volt;
+    
+
+    // calculate time since last current read
+    uint32_t tnow = AP_HAL::micros();
+    float dt = tnow - _state.last_time_micros;
+
+    // update total current drawn since startup
+    if (_state.last_time_micros != 0 && dt < 2000000.0f) {
+        // .0002778 is 1/3600 (conversion to hours)
+        float mah = _state.current_amps * dt * 0.0000002778f;
+        _state.consumed_mah += mah;
+        _state.consumed_wh  += 0.001f * mah * _state.voltage;
+    }
+
+    GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "Curr Pin : %d, raw_Value : %f final Value : %f _curr_amp_offset : %f  _curr_amp_per_volt : %f ",
+                                                curr_buff_pt, 
+                                                _data_to_volt(_analog_data[curr_buff_pt].data),
+                                                _state.current_amps,
+                                                _curr_amp_offset.get(),
+                                                _curr_amp_per_volt.get()
+                                                );         
+    // record time
+    _state.last_time_micros = tnow;
+}
+
+/**
+ * @brief Timmer function to read raw adc
+ * 
+ */
+void AP_BattMonitor_AD7091R5::_timer(){
+    _read_adc();
```

Comment:
Maybe just remove this wrapper and just pass the _read_adc function to be called instead (or combine them).